internal/providers/zvm/zvm.go

```diff
@@ -0,0 +1,102 @@
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// The zVM provider fetches a local configuration from the virtual unit
+// record devices.
+
+package zvm
+
+import (
+	"io/ioutil"
+	"os/exec"
+	"strings"
+
+	"github.com/coreos/ignition/v2/config/shared/errors"
+	"github.com/coreos/ignition/v2/config/v3_1_experimental/types"
+	"github.com/coreos/ignition/v2/internal/log"
+	"github.com/coreos/ignition/v2/internal/providers/util"
+	"github.com/coreos/ignition/v2/internal/resource"
+	"github.com/coreos/vcontext/report"
+)
+
+const readerDevice string = "000c"
+const configFile string = "config.ign"
+
+func FetchConfig(f *resource.Fetcher) (types.Config, report.Report, error) {
+	// Fetch config files directly from reader device.
+	_, err := f.Logger.LogCmd(exec.Command("modprobe", "vmur"), "Loading zVM control program module")
+	if err != nil {
+		f.Logger.Err("Couldn't install vmur module: %v", err)
+		return types.Config{}, report.Report{}, err
+	}
+	// Online the reader device.
+	logger := f.Logger
+	rc := onlineDevice(logger)
+	if !rc {
+		return types.Config{}, report.Report{}, errors.ErrEmpty
+	}
+	// Read files from the z/VM reader queue.
+	var comm string = "vmur li | tail -n +2 | cut -c 1-80 | awk '{print $1,$2,$3,$10,$11}'"
+	cmd := exec.Command("/bin/sh", "-c", comm)
+	out, err := cmd.Output()
+	if err != nil {
+		f.Logger.Err("Can not get reader device: %v", err)
+	}
+	items := strings.Split(string(out), "\n")
+	for _, records := range items {
+		record := strings.Split(records, " ")
+		if len(record) == 1 {
+			break
+		}
+		spoolid := record[1]
+		filename := record[3]
+		types := record[4]
+		var file string
+		if types == "" {
+			file = filename
+		} else {
+			file = filename + "." + types
+		}
+		// Receive the spool file.
+		if filename == "ignition" {
+			_, err := f.Logger.LogCmd(exec.Command("vmur", "re", "-f", spoolid, file), "Receive the spool file")
+			if err == nil {
+				f.Logger.Info("using config file at %q", file)
+				_, err := f.Logger.LogCmd(exec.Command("tar", "-xzvf", file), "Extract the spool file")
+				rawConfig, err := ioutil.ReadFile(configFile)
+				if err != nil {
+					f.Logger.Err("Couldn't read config %q: %v", filename, err)
+					break
+				}
+				return util.ParseConfig(f.Logger, rawConfig)
+			}
+		}
+		_ = file
+	}
+	return types.Config{}, report.Report{}, errors.ErrEmpty
+}
+
+func onlineDevice(logger *log.Logger) bool {
+	_, err := logger.LogCmd(exec.Command("chccwdev", "-e", readerDevice), "Brings a Linux device online")
+	if err != nil {
+		_, err = logger.LogCmd(exec.Command("cio_ignore", "-r", readerDevice), "Expose reader device")
```

Comment:
This needs an `if err != nil` check too, otherwise the next line will overwrite it.